### ISSUE TITLE
Add color-no-out-gamut-range

### ISSUE THREAD

**Issue by ai** [stylelint/stylelint#5905]
### What is the problem you're trying to solve?

In new Lab/OKLab color, it is easy to move color to P3 space by mistake (for instance, by round floats after HSL→LCH convert).

P3 color could looks very different on sRGB monitor causing a11y issues.

### What solution would you like to see?

It will very useful if Stylelint will warn you, that it is a P3 color, and it is not in [`color-gamut` media](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/color-gamut).

**Comment by ybiquitous** [comment]
@ai Thanks for the suggestion and for using the template. Sounds good to me. 👍🏼 

I cannot find a spec for `oklch()` and `oklab()`, so could you please provide where the spec is?

**Comment by ai** [comment]
We may need option to change safe gamut: `srgb` (default), `p3`, `rec2020`.

Note, that this rule not about feature support. Even when all browsers will support `lch()` and P3 colors, we will still have screens that doesn’t support P3 colors. Most of LCD screens doesn’y support it right now.

> I cannot find a spec for oklch() and oklab(), so could you please provide where the spec is?

https://www.w3.org/TR/css-color-4/#resolving-oklab-oklch-values

Lab/LCH has an issue: if you change `C` component, it could accidentally change hue in blue part of spectrum. OKLab/OKLCH fixes this issue.

See also: https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-oklab-function

> Can you also provide some examples, please?

Allow:

```css
a {
  color: lch(50% 0 0);
  color: lch(48.8% 82.3 283.1);
}

@media (color-gamut: p3) {
  a {
    color: lch(48% 82 283); // P3 color
  }
}
```

Disallow:

```css
a {
  color: lch(48% 82 283);
}
```

You can use [LCH Colour Picker](https://css.land/lch/) or [Huetone](https://huetone.ardov.me/) to check manually that color is inside sRGB gamut or go to P3.

To detect it in rule we can use [color.js](https://colorjs.io/). See also [warning](https://github.com/csstools/postcss-plugins/blob/bd275c6e43c1f276c5afcdad6caae16078080585/plugins/postcss-lab-function/src/on-css-function.ts#L125) in [postcss-lab-function] polyfill.

**Comment by jeddy3** [comment]
Thanks for sharing more information and examples.

- Name: `color-no-out-gamut-range`
- Primary option: `true`
- Secondary options: None. (Later down the line, we can add a `range: "srgb" (default)|"p3"|"rec2020"` secondary option to make the rule more permissive.)
- Autofixable: No
- Message: `Unexpected color out of gamut range`
- Description: "Disallow out of gamut range colors"
- Section: "Avoid errors" -> "Colors"


(Suggested blueprint: [color-no-invalid-hex](https://stylelint.io/user-guide/rules/list/color-no-invalid-hex)).

We'll need to document the limitation that conditional imports won't be respected, e.g.:

```html
<link href="p3-custom-props.css" rel="stylesheet" media="(color-gamut: p3)">
```

```css
@import url('p3-custom-props.css') (color-gamut: p3);
```

As this rule touches on very new stuff, I think details will emerge if a PoC pull request is created. We can decide then if this is better suited as a built-in rule or plugin. It'd be easy to port the code from the pull request to a plugin, if we end up going that way.

> To detect it in rule we can use [color.js](https://colorjs.io/).

If we use this package, we should send a PR upstream to use `files: []` is the `package.json` as it's heavy.

I've labelled the issue as ready to implement.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

---

Further reading resources:

- https://css-tricks.com/new-css-color-features-preview/
- https://twitter.com/css_tools_/status/1493884620747816965
- https://twitter.com/sitnikcode/status/1493208704010833921

**Comment by ai** [comment]
We can also use @omgovich `colord` when they will add [OKLab support](https://github.com/omgovich/colord/issues/87).

`colord` is much [smaller](https://packagephobia.com/result?p=colord) than `color.js`.

**Comment by dkryaklin** [comment]
FYI There is an open PR for OKLab support in `colord` library
https://github.com/omgovich/colord/pull/92

**Comment by fpetrakov** [comment]
Hey, I really want to work on this new rule and I want to clarify some things beforehand. Correct me if I'm wrong.

1. This rule should check colors declared with `lch(), lab(), oklch(), oklab()`.
2. If declared color is out of sRGB range, we check if it's wrapped in `@media (color-gamut: p3) {}`.
3. If it's not wrapped a media query then we should throw warning.

**Comment by jeddy3** [comment]
Available as plugin: https://www.npmjs.com/package/stylelint-gamut
